frontend/dockerui/build.go

```diff
@@ -0,0 +1,103 @@
+package dockerui
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/containerd/containerd/platforms"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	"github.com/moby/buildkit/exporter/containerimage/image"
+	"github.com/moby/buildkit/frontend/gateway/client"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+type BuildFunc func(ctx context.Context, platform *ocispecs.Platform, idx int) (client.Reference, *image.Image, error)
+
+func (bc *Client) Build(ctx context.Context, fn BuildFunc) (*ResultBuilder, error) {
+	res := client.NewResult()
+
+	targets := make([]*ocispecs.Platform, 0, len(bc.TargetPlatforms))
+	for _, p := range bc.TargetPlatforms {
+		p := p
+		targets = append(targets, &p)
+	}
+	if len(targets) == 0 {
+		targets = append(targets, nil)
+	}
+	expPlatforms := &exptypes.Platforms{
+		Platforms: make([]exptypes.Platform, len(targets)),
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+
+	for i, tp := range targets {
+		i, tp := i, tp
+		eg.Go(func() error {
+			ref, img, err := fn(ctx, tp, i)
```

Comment:
What do you suggest for the solution for the `Warn` issue then?

Reply:
Yeah, not sure :thinking: I tried playing around with maybe doing platform matching here, but it's way more hacky than just having the index parameter.

I think it's fine as is, but we should watch this interface to avoid changing it that much in the future, if we want frontend authors to use it and be able to easily upgrade between buildkit versions.